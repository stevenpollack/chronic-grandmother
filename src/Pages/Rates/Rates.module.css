.container {
    display: flex;
    justify-content: flex-start;
    align-items: center;
    flex-direction: column;
    width: 100%;
    height: 100%;
}
.content {
    margin-top: 100px;
}

.heading {
    color: #142557;
    font-weight: 700;
    font-size: 24px;
    line-height: 33px;
    width: 100%;
    margin-bottom: 25px;
}

.rowWrapper {
    display: flex;
    flex-direction: row;
}
.exchangeWrapper {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: flex-end;
    margin: 0px 30px;
}

.transferIcon {
    margin-bottom: 18px;
}
.transferIcon > svg > path {
    fill: #142557;
}
.rate {
    font-size: 18px;
    color: #142557;
    opacity: 0.7;
}

.flag {
    margin-right: 10px;
    width: 26px;
    height: 20px;
    margin-left: 5px;
}

.loaderWrapper {
    display: flex;
    justify-content: center;
    margin-top: 20px;
}
.slow {
    transition: all 1s cubic-bezier(0.55, 0.055, 0.675, 0.19) !important;
}
