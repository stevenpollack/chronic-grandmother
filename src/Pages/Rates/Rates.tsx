import { useState } from 'react';
import DropDown from '../../Components/DropDown';
import ProgressBar from '../../Components/ProgressBar';
import Loader from '../../Components/Loader';

import { useAnimationFrame } from '../../Hooks/useAnimationFrame';

import classes from './Rates.module.css';

import CountryData from '../../Libs/Countries.json';
import countryToCurrency from '../../Libs/CountryCurrency.json';

let countries = CountryData.CountryCodes;

const Rates = () => {
    const [fromCurrency, setFromCurrency] = useState('AU');
    const [toCurrency, setToCurrency] = useState('US');

    const [exchangeRate, setExchangeRate] = useState(0.7456);
    const [progression, setProgression] = useState(0);
    const [loading, setLoading] = useState(false);

    const Flag = ({ code }: { code: string }) => (
        <img alt={code || ''} src={`/img/flags/${code || ''}.svg`} width="20px" className={classes.flag} />
    );

    const fetchData = async () => {
        if (!loading) {
            setLoading(true);

            await new Promise((resolve) => setTimeout(resolve, 2000));

            setLoading(false);
        }
    };

    // Demo progress bar moving :)
    useAnimationFrame(!loading, (deltaTime) => {
        setProgression((prevState) => {
            if (prevState > 0.998) {
                fetchData();
                return 0;
            }
            return (prevState + deltaTime * 0.0001) % 1;
        });
    });

    return (
        <div className={classes.container}>
            <div className={classes.content}>
                <div className={classes.heading}>Currency Conversion</div>

                <div className={classes.rowWrapper}>
                    <div>
                        <DropDown
                            leftIcon={<Flag code={fromCurrency} />}
                            label={'From'}
                            selected={countryToCurrency[fromCurrency as keyof typeof countryToCurrency]}
                            options={countries.map(({ code }) => ({
                                option: countryToCurrency[code as keyof typeof countryToCurrency],
                                key: code,
                                icon: <Flag code={code} />,
                            }))}
                            setSelected={(key: string) => {
                                setFromCurrency(key);
                            }}
                            style={{ marginRight: '20px' }}
                        />
                    </div>

                    <div className={classes.exchangeWrapper}>
                        <div className={classes.transferIcon}>
                            <img src="/img/icons/Transfer.svg" alt="Transfer icon" />
                        </div>

                        <div className={classes.rate}>{exchangeRate}</div>
                    </div>

                    <div>
                        <DropDown
                            leftIcon={<Flag code={toCurrency} />}
                            label={'To'}
                            selected={countryToCurrency[toCurrency as keyof typeof countryToCurrency]}
                            options={countries.map(({ code }) => ({
                                option: countryToCurrency[code as keyof typeof countryToCurrency],
                                key: code,
                                icon: <Flag code={code} />,
                            }))}
                            setSelected={(key: string) => {
                                setToCurrency(key);
                            }}
                            style={{ marginLeft: '20px' }}
                        />
                    </div>
                </div>

                <ProgressBar
                    progress={progression}
                    animationClass={loading ? classes.slow : ''}
                    style={{ marginTop: '20px' }}
                />

                {loading && (
                    <div className={classes.loaderWrapper}>
                        <Loader width={'25px'} height={'25px'} />
                    </div>
                )}
            </div>
        </div>
    );
};

export default Rates;
