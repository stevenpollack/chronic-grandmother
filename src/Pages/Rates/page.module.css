.container {
  display: flex;
  justify-content: flex-start;
  align-items: center;
  flex-direction: column;
  width: 100%;
  height: 100%;
}
.content {
  margin-top: 100px;
}

.heading {
  color: #142557;
  font-weight: 700;
  font-size: 24px;
  line-height: 33px;
  width: 100%;
  margin-bottom: 25px;
}