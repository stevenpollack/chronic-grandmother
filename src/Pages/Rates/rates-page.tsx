import { useLocation } from "react-router-dom";

import CurrencyConverter from "../../Components/CurrencyConverter/currency-converter";
import classes from "./page.module.css";

export default function RatesPage() {
  const { search } = useLocation();
  const params = new URLSearchParams(search);

  // these parameters are for debugging purposes
  const margin = Number(params.get("margin") ?? 0.005);
  const refreshRate = Number(params.get("refreshRate") ?? 10_000);
  const maxRetries = Number(params.get("maxRetries") ?? 3);

  return (
    <div className={classes.container}>
      <div className={classes.content}>
        <h1 className={classes.heading}>Currency Conversion</h1>

        <CurrencyConverter
          margin={margin}
          refreshRate={refreshRate}
          maxRetries={maxRetries}
        />
      </div>
    </div>
  );
}
