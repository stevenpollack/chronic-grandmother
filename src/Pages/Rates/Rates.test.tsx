import { server } from "../../mocks/node";
import Rates, { API_URL, ApiResponse } from "./Rates";
import { render, waitFor } from "@testing-library/react";
import { http, HttpResponse } from "msw";
import userEvent from "@testing-library/user-event";

describe("Rates", () => {
  const RETAIL_RATE = 0.6543;
  const apiPath = API_URL.toString();
  const spy = vi.fn();

  const successfulResponse = http.get(apiPath, ({ request }) => {
    const requestURL = request.url;
    const url = new URL(requestURL);
    spy(requestURL);
    return HttpResponse.json<ApiResponse>({
      buyCurrency: url.searchParams.get("buyCurrency") || "",
      createdAt: "2021-01-01",
      id: "1",
      indicative: false,
      retailRate: RETAIL_RATE,
      sellCurrency: url.searchParams.get("sellCurrency") || "",
      validUntil: "2021-01-01",
      wholesaleRate: 1.0,
    });
  });

  const errorResponse = http.get(apiPath, () => {
    return HttpResponse.json(
      {
        error: "API error",
      },
      { status: 500 }
    );
  });

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("calls the rates API on render", async () => {
    server.use(successfulResponse);
    render(<Rates refreshRate={10} />);
    await waitFor(() => {
      expect(spy).toHaveBeenCalled();
    });
  });

  it("renders correctly with default props", async () => {
    server.use(successfulResponse);
    const { asFragment, getByTestId } = render(<Rates refreshRate={10} />);

    await waitFor(() => {
      expect(getByTestId("exchange-rate")).toHaveTextContent(
        RETAIL_RATE.toString()
      );
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders an error message when the API call fails", async () => {
    server.use(errorResponse);
    const { getByTestId } = render(<Rates refreshRate={10} maxRetries={1} />);

    await waitFor(() => {
      expect(getByTestId("error-message")).toBeVisible();
    });
  });

  it("renders a functional retry button when API call fails", async () => {
    server.use(errorResponse);
    const { getByTestId } = render(<Rates refreshRate={250} maxRetries={1} />);
    const user = userEvent.setup();

    await waitFor(() => {
      expect(getByTestId("retry-button")).toBeInTheDocument();
      expect(getByTestId("retry-button")).toBeEnabled();
    });

    // Retry button should trigger a new API call
    server.use(successfulResponse);
    await user.click(getByTestId("retry-button"));
    await waitFor(() => {
      expect(successfulResponse.isUsed).toBe(true);
    });
  });
});
