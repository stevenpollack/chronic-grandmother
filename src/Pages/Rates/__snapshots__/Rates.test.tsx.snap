// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Rates > renders correctly with default props 1`] = `
<DocumentFragment>
  <div
    class="_container_3590bb"
  >
    <div
      class="_content_3590bb"
    >
      <h1
        class="_heading_3590bb"
      >
        Currency Conversion
      </h1>
      <div
        class="_rowWrapper_3590bb"
      >
        <div
          class="_dropdownContainer_3590bb _fromDropdown_3590bb"
        >
          <div
            class="_container_5d938f undefined"
          >
            <label
              class="_label_5d938f"
              id="From-label"
            >
              From
            </label>
            <button
              aria-describedby="From-value"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="From-label"
              class="_dropdown_5d938f"
              data-testid="dropdown-button"
            >
              <img
                alt="AU"
                class="_flag_479884"
                data-testid="AU-flag-icon"
                src="/img/flags/AU.svg"
                width="20px"
              />
              <span
                class="_dropdownText_5d938f"
                data-testid="From-value"
              >
                AUD
              </span>
              <div
                aria-hidden="true"
                class="_dropdownToggle_5d938f"
              >
                <svg
                  fill="none"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16.2 5.3894H7.8C6.2536 5.3894 5 6.64301 5 8.1894V16.5894C5 18.1358 6.2536 19.3894 7.8 19.3894H16.2C17.7464 19.3894 19 18.1358 19 16.5894V8.1894C19 6.64301 17.7464 5.3894 16.2 5.3894Z"
                    stroke="#E5E5E5"
                    stroke-width="1.03427"
                  />
                  <path
                    class="_toggleArrow_5d938f "
                    clip-rule="evenodd"
                    d="M9.20711 11.3894C8.76165 11.3894 8.53857 11.928 8.85355 12.243L11.6464 15.0359C11.8417 15.2311 12.1583 15.2311 12.3536 15.0358L15.1464 12.243C15.4614 11.928 15.2383 11.3894 14.7929 11.3894H9.20711Z"
                    fill="#6D7587"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
        <div
          class="_exchangeWrapper_3590bb"
        >
          <div
            class="_transferIcon_3590bb"
          >
            <img
              alt="Transfer icon"
              src="/img/icons/Transfer.svg"
            />
          </div>
        </div>
        <div
          class="_dropdownContainer_3590bb _toDropdown_3590bb"
        >
          <div
            class="_container_5d938f undefined"
          >
            <label
              class="_label_5d938f"
              id="To-label"
            >
              To
            </label>
            <button
              aria-describedby="To-value"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="To-label"
              class="_dropdown_5d938f"
              data-testid="dropdown-button"
            >
              <img
                alt="US"
                class="_flag_479884"
                data-testid="US-flag-icon"
                src="/img/flags/US.svg"
                width="20px"
              />
              <span
                class="_dropdownText_5d938f"
                data-testid="To-value"
              >
                USD
              </span>
              <div
                aria-hidden="true"
                class="_dropdownToggle_5d938f"
              >
                <svg
                  fill="none"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16.2 5.3894H7.8C6.2536 5.3894 5 6.64301 5 8.1894V16.5894C5 18.1358 6.2536 19.3894 7.8 19.3894H16.2C17.7464 19.3894 19 18.1358 19 16.5894V8.1894C19 6.64301 17.7464 5.3894 16.2 5.3894Z"
                    stroke="#E5E5E5"
                    stroke-width="1.03427"
                  />
                  <path
                    class="_toggleArrow_5d938f "
                    clip-rule="evenodd"
                    d="M9.20711 11.3894C8.76165 11.3894 8.53857 11.928 8.85355 12.243L11.6464 15.0359C11.8417 15.2311 12.1583 15.2311 12.3536 15.0358L15.1464 12.243C15.4614 11.928 15.2383 11.3894 14.7929 11.3894H9.20711Z"
                    fill="#6D7587"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
        <div
          class="_inputContainer_3590bb _fromInput_3590bb"
        >
          <label
            class="sr-only"
            for="amount-input"
          >
            Amount to convert
          </label>
          <input
            aria-describedby="conversion-results"
            aria-label="Amount to convert"
            class="_amountInput_3590bb"
            id="amount-input"
            placeholder="0.00"
            type="number"
            value=""
          />
        </div>
        <div
          class="_exchangeWrapperBottom_3590bb"
        >
          <div
            class="_rate_3590bb"
            data-testid="exchange-rate"
          >
            0.6543
          </div>
        </div>
        <div
          aria-live="polite"
          class="_amountWrapper_3590bb"
          id="conversion-results"
        >
          <span
            aria-label="OFX conversion result: 0.00 USD"
            class="_rate_3590bb"
          >
            OFX rate: 0.00
          </span>
          <span
            aria-label="True conversion result: 0.00 USD"
            class="_rate_3590bb"
          >
            True rate: 0.00
          </span>
        </div>
      </div>
      <div
        class="_base_a40bf7"
        style="margin-top: 20px;"
      >
        <div
          class="_progress_a40bf7 "
          style="transform: translate(-0px);"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
